compendium/DeclarativeServices/src/manager/ConfigurationNotifier.hpp

```diff
@@ -0,0 +1,125 @@
+/*=============================================================================
+
+ Library: CppMicroServices
+
+ Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+ file at the top-level directory of this distribution and at
+ https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+ =============================================================================*/
+
+#ifndef __CPPMICROSERVICES_SCRIMPL_CONFIGURATIONNOTIFIER_HPP__
+#define __CPPMICROSERVICES_SCRIMPL_CONFIGURATIONNOTIFIER_HPP__
+
+#include "../SCRLogger.hpp"
+#include "ConcurrencyUtil.hpp"
+#include "cppmicroservices/cm/ConfigurationListener.hpp"
+//#include "ComponentConfigurationImpl.hpp"
+
+namespace cppmicroservices {
+namespace scrimpl {
+class ComponentConfigurationImpl;
+
+/** ConfigChangeNotification
+     * This class is used by ConfigurationListener to notify ComponentConfigurationImpl
+     * about changes to Configuration Objects.
+     */
+struct ConfigChangeNotification final
+{
+  ConfigChangeNotification(
+    const std::string pid,
+    std::shared_ptr<cppmicroservices::AnyMap> properties,
+    const cppmicroservices::service::cm::ConfigurationEventType evt)
+    : pid(std::move(pid))
+    , event(std::move(evt))
+    , newProperties(properties)
+  {}
+
+  const std::string pid;
+  const cppmicroservices::service::cm::ConfigurationEventType event;
+  std::shared_ptr<cppmicroservices::AnyMap> newProperties;
+};
+
+struct Listener final
+{
+  Listener(std::function<void(const ConfigChangeNotification&)> notify,
+           std::shared_ptr<ComponentConfigurationImpl> mgr)
+    : notify(std::move(notify))
+    , mgr(std::move(mgr))
+  {}
+
+  std::function<void(const ConfigChangeNotification&)> notify;
+  std::shared_ptr<ComponentConfigurationImpl> mgr;
+};
+
+class ConfigurationNotifier final
+{
+
+public:
+  /**
+   * @throws std::invalid_argument exception if any of the params is a nullptr 
+   */
+  ConfigurationNotifier(
+    const cppmicroservices::BundleContext context,
```

Comment:
Jeff: check consistency of this bundle context (choose one strategy across other classes), probably const ref